File: reverse/reverse.c
#include <stdlib.h>

__attribute__((export_name("allocate_reverse")))
void ** allocate_reverse(uint32_t length) {
  unsigned char *original = malloc(length);
  unsigned char *reversed = malloc(length);
  uint32_t *stored_length = malloc(sizeof(uint32_t));

  *stored_length = length;

  void **pointers = malloc(sizeof(size_t) * 3);
  pointers[0] = stored_length;
  pointers[1] = original;
  pointers[2] = reversed;

  return pointers;
}

__attribute__((export_name("reverse")))
int reverse(void **pointers) {
  const uint32_t length = *(uint32_t *)pointers[0];
  const unsigned char *original = (unsigned char *)pointers[1];
  unsigned char *reversed = (unsigned char *)pointers[2];

  for (int i = 0; i < length; i++) {
    int position = length - i - 1;
    reversed[position] = original[i];
  }

  return 0;
}

__attribute__((export_name("free_reverse")))
int free_reverse(void **pointers) {
  void *length = pointers[0];
  void *original = pointers[1];
  void *reversed = pointers[2];

  free(length);
  free(original);
  free(reversed);

  return 0;
}

int main() {
  return 0;
}

File: reverse/reverse.ts
type Exports = {
  memory: WebAssembly.Memory
  allocate_reverse(length: number): number
  reverse(pointers: number): number
  free_reverse(pointers: number): number
}

const encoder = new TextEncoder()
const decoder = new TextDecoder()

const originalString = Deno.args.join(' ')
const originalBytes = encoder.encode(originalString)

const code = await Deno.readFile('reverse.wasm')
const { instance } = await WebAssembly.instantiate(code)
const ex = instance.exports as Exports
const memory = ex.memory

const pointers = ex.allocate_reverse(originalBytes.byteLength)
const view = new DataView(memory.buffer, pointers, 12) // 3 32bit integers is 3 * 4
const storedLengthPointer = view.getUint32(0, true)
const originalPointer = view.getUint32(4, true)
const reversedPointer = view.getUint32(8, true)
const storedLength = (new DataView(memory.buffer, storedLengthPointer, 4)).getUint32(0, true)

if (storedLength !== originalBytes.byteLength) {
  console.error('Wrong storedLength, something has gone wrong')
  Deno.exit(1)
}

const wasmOriginalBytes = new Uint8Array(memory.buffer, originalPointer, storedLength)
wasmOriginalBytes.set(originalBytes)

if (ex.reverse(pointers) !== 0) {
  console.error('reverse failed')
  Deno.exit(1)
}

const reversedBytes = Uint8Array.from(new Uint8Array(memory.buffer, reversedPointer, storedLength))

if (ex.free_reverse(pointers) !== 0) {
  console.error('free_reverse failed')
  Deno.exit(1)
}

console.log(`Original: ${originalString}`)
console.log(`Reversed: ${decoder.decode(reversedBytes)}`)
